### PR TITLE
Add support of varchar(x) to presto scalar functions and operators.

### PR DESCRIPTION
**Pull request by fiedukow**
@losipiuk @arhimondr

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ColorFunctions.java [hunk 1]
```diff
@@ -149,8 +152,9 @@ public static long color(@SqlType(StandardTypes.DOUBLE) double fraction, @SqlTyp
     }
 
     @ScalarFunction
-    @SqlType(StandardTypes.VARCHAR)
-    public static Slice render(@SqlType(StandardTypes.VARCHAR) Slice value, @SqlType(ColorType.NAME) long color)
+    @LiteralParameters("x")
+    @SqlType(VarcharType.VARCHAR_MAX_LENGTH)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ColorFunctions.java]: I think it is possible to determine more precise output type. Ansi encoded color is fixed length value. So the result type should be `varchar(x + ansi_encoded_color_length_in_characters)`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java [hunk 2]
```diff
@@ -167,7 +169,7 @@ public static double toUnixTimeFromTimestampWithTimeZone(@SqlType(StandardTypes.
     }
 
     @ScalarFunction("to_iso8601")
-    @SqlType(StandardTypes.VARCHAR)
+    @SqlType(VarcharType.VARCHAR_MAX_LENGTH)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java]: I think we can determine max characters count for date or timestamp in `iso8601` format. Can't we?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java [hunk 3]
```diff
@@ -515,18 +539,20 @@ private static Slice formatDatetime(ISOChronology chronology, Locale locale, lon
     }
 
     @ScalarFunction
-    @SqlType(StandardTypes.VARCHAR)
-    public static Slice dateFormat(ConnectorSession session, @SqlType(StandardTypes.TIMESTAMP) long timestamp, @SqlType(StandardTypes.VARCHAR) Slice formatString)
+    @LiteralParameters("x")
+    @SqlType(VarcharType.VARCHAR_MAX_LENGTH)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java:543]: Shouldn't the length be the same as a format string length?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java [hunk 4]
```diff
@@ -541,8 +567,9 @@ private static Slice dateFormat(ISOChronology chronology, Locale locale, long ti
     }
 
     @ScalarFunction
+    @LiteralParameters("x")
     @SqlType(StandardTypes.TIMESTAMP)
-    public static long dateParse(ConnectorSession session, @SqlType(StandardTypes.VARCHAR) Slice dateTime, @SqlType(StandardTypes.VARCHAR) Slice formatString)
+    public static long dateParse(ConnectorSession session, @SqlType("varchar(x)") Slice dateTime, @SqlType(StandardTypes.VARCHAR) Slice formatString)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java]: s/`StandardTypes.VARCHAR`/`varchar(y)`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java [hunk 5]
```diff
@@ -76,16 +80,18 @@ public static boolean regexpLike(@SqlType(StandardTypes.VARCHAR) Slice source, @
...
     {
         return regexpReplace(source, pattern, Slices.EMPTY_SLICE);
     }
 
     @Description("replaces substrings matching a regular expression by given string")
     @ScalarFunction
-    @SqlType(StandardTypes.VARCHAR)
-    public static Slice regexpReplace(@SqlType(StandardTypes.VARCHAR) Slice source, @SqlType(RegexpType.NAME) Regex pattern, @SqlType(StandardTypes.VARCHAR) Slice replacement)
+    @LiteralParameters({"x", "y"})
+    @SqlType(VarcharType.VARCHAR_MAX_LENGTH)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java]: Can we put `varchar(x * y)` here?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java [hunk 6]
```diff
@@ -195,16 +201,18 @@ private static void appendReplacement(SliceOutput result, Slice source, Regex pa
 
     @Description("string(s) extracted using the given pattern")
     @ScalarFunction
+    @LiteralParameters("x")
     @SqlType("array(varchar)")
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java]: I think it is better to specify varchar(MAX_LENGTH) explicitly here. I don't know whether we are going to support unparametrized `varchar` signature for `varchar(x)` type.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java [hunk 7]
```diff
@@ -195,16 +201,18 @@ private static void appendReplacement(SliceOutput result, Slice source, Regex pa
...
-    public static Block regexpExtractAll(@SqlType(StandardTypes.VARCHAR) Slice source, @SqlType(RegexpType.NAME) Regex pattern)
+    public static Block regexpExtractAll(@SqlType("varchar(x)") Slice source, @SqlType(RegexpType.NAME) Regex pattern)
     {
         return regexpExtractAll(source, pattern, 0);
     }
 
     @Description("group(s) extracted using the given pattern")
     @ScalarFunction
+    @LiteralParameters("x")
     @SqlType("array(varchar)")
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java]: ditto

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java [hunk 8]
```diff
@@ -273,9 +283,10 @@ public static Slice regexpExtract(@SqlType(StandardTypes.VARCHAR) Slice source,
     }
 
     @ScalarFunction
+    @LiteralParameters("x")
     @Description("returns array of strings split by pattern")
     @SqlType("array(varchar)")
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java]: ditto

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java (hunk 6 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java]: Actually i think it is even possible to put there `varchar(x)`, by analogy to `split` function

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/UrlFunctions.java [hunk 9]
```diff
@@ -136,17 +145,20 @@ public static Slice urlExtractParameter(@SqlType(StandardTypes.VARCHAR) Slice ur
 
     @Description("escape a string for use in URL query parameter names and values")
     @ScalarFunction
-    @SqlType(StandardTypes.VARCHAR)
-    public static Slice urlEncode(@SqlType(StandardTypes.VARCHAR) Slice value)
+    @LiteralParameters({"x", "y"})
+    @Constraint(variable = "y", expression = "min(2147483647, x*3)")
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/UrlFunctions.java]: UTF characters can be 4 bytes long. And in order to represent UTF in URL octal notation you need 3 characters per byte e.g. `encodeURIComponent("Ж") -> %D0%96`, `encodeURIComponent("𩸽") -> %F0%A9%B8%BD`. So actually result type should be calculated as `min(2147483647, x*12)`

File: presto-main/src/main/java/com/facebook/presto/type/BigintOperators.java [hunk 10]
```diff
@@ -195,8 +195,10 @@ public static double castToDouble(@SqlType(StandardTypes.BIGINT) long value)
         return value;
     }
 
+    // x <= 20
     @ScalarOperator(CAST)
-    @SqlType(StandardTypes.VARCHAR)
+    @LiteralParameters("x")
+    @SqlType("varchar(x)")
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/type/BigintOperators.java:200]: For all the cast operators we should check that the result doesn't exceed the result type length. And based on the cast type either to trim the result value to the result type, or throw the exception.

**Comment by arhimondr** [comment]
lgtm % comments

**Comment by arhimondr** [comment]
FWIW, I think this patch can be separated into smaller commits. Like `Refactor varchar functions in MathFunction`, `Refactor varchar functions in PathFunctions`, etc. No need to do it now. It is still pretty easy to follow, and artificial separation now might be painful. 

There are basically 2 reasons for doing so:
- It is easier to follow smaller commits
- You will be higher in contributors rating of Presto, cause the rating is sorted by the commits count =)

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java (hunk 3 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/DateTimeFunctions.java:543]: I don't think so.

```
    @Test
    public void testDateToTimestampCoercing()
    {
        assertFunction("date_format(DATE '2013-10-27', '%Y-%m-%d %H:%i:%s')", VARCHAR, "2013-10-27 00:00:00");
        //...
    }
```

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/UrlFunctions.java (hunk 9 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/UrlFunctions.java]: done, also covered that in tests.

File: presto-main/src/main/java/com/facebook/presto/type/BigintOperators.java (hunk 10 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/type/BigintOperators.java:200]: I've marked this as FIXME.
There is a need to implement possibility of passing literal long parameters as function parameter as prerequirement for fixing this.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ColorFunctions.java (hunk 1 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ColorFunctions.java]: true, done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java (hunk 5 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/RegexpFunctions.java]: I hope that there is no crazy regex that exeeds that ;-)

**Comment by fiedukow** [comment]
I think I'm done with fixes except of casts operators out of range support which requires another task to be completed earlier.

**Pull request closed without merging**